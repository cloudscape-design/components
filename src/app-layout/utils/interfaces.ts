// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
import React from 'react';

export interface SizeControlProps {
  position: 'side' | 'bottom';
  panelRef?: React.RefObject<HTMLDivElement>;
  handleRef?: React.RefObject<HTMLDivElement>;
  onResize: (newSize: number) => void;
  hasTransitions?: boolean;
}
