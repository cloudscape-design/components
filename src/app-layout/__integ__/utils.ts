// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

export const testIf = (condition: boolean) => (condition ? test : test.skip);

export type Theme = 'classic' | 'refresh' | 'refresh-toolbar';

export function getUrlParams(theme: Theme) {
  const params = new URLSearchParams({
    visualRefresh: `${theme.startsWith('refresh')}`,
    appLayoutToolbar: `${theme === 'refresh-toolbar'}`,
  });
  return params.toString();
}
