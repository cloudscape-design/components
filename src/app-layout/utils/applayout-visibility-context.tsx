// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
import { createContext } from 'react';

export const AppLayoutVisibilityContext = createContext<boolean>(true);
