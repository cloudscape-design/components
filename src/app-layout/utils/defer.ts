// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
export class Deferred<T> {
  private readonly _promise: Promise<T>;
  private _resolve?: (value?: any) => void;
  private _reject?: (reason?: any) => void;

  constructor() {
    this._promise = new Promise<T>((resolve, reject) => {
      this._resolve = resolve;
      this._reject = reject;
    });
  }

  get promise(): Promise<T> {
    return this._promise;
  }

  resolve = (value?: T | PromiseLike<T>): void => {
    this._resolve?.(value);
  };

  reject = (reason?: any): void => {
    this._reject?.(reason);
  };
}
