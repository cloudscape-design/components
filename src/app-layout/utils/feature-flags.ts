// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
import { getGlobalFlag } from '@cloudscape-design/component-toolkit/internal';

import { useVisualRefresh } from '../../internal/hooks/use-visual-mode';

export const useAppLayoutToolbarEnabled = () => {
  const isRefresh = useVisualRefresh();
  return isRefresh && (getGlobalFlag('appLayoutWidget') || getGlobalFlag('appLayoutToolbar'));
};
