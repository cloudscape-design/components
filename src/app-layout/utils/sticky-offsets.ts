// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import globalVars from '../../internal/styles/global-vars';

export function getStickyOffsetVars(
  headerHeight: number,
  footerHeight: number,
  stickyNotificationsHeight: string,
  mobileToolbarHeight: string,
  disableBodyScroll: boolean,
  isMobile: boolean,
  extraTopOffset?: string
) {
  return {
    [globalVars.stickyVerticalTopOffset]: `calc(${!disableBodyScroll ? headerHeight : 0}px + ${
      isMobile ? mobileToolbarHeight : stickyNotificationsHeight
    } + ${extraTopOffset ?? '0px'})`,
    [globalVars.stickyVerticalBottomOffset]: `${!disableBodyScroll ? footerHeight : 0}px`,
  };
}
