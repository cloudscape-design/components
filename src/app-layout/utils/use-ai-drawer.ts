// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
import { useEffect, useRef, useState } from 'react';

import { useStableCallback } from '@cloudscape-design/component-toolkit/internal';

import { fireNonCancelableEvent } from '../../internal/events';
import { metrics } from '../../internal/metrics';
import { AppLayoutMessage, DrawerPayload as RuntimeAiDrawerConfig } from '../../internal/plugins/widget/interfaces';
import { getAppLayoutInitialState, registerAppLayoutHandler } from '../../internal/plugins/widget/internal';
import { assertNever } from '../../internal/types';
import { mapRuntimeConfigToAiDrawer } from '../runtime-drawer';

export interface OnChangeParams {
  initiatedByUserAction: boolean;
}

const DEFAULT_ON_CHANGE_PARAMS = { initiatedByUserAction: true };

function useRuntimeAiDrawer(
  isEnabled: boolean,
  activeAiDrawerId: string | null,
  onActiveAiDrawerChange: (newDrawerId: string | null, { initiatedByUserAction }: OnChangeParams) => void,
  onActiveAiDrawerResize: (size: number) => void
) {
  const [aiDrawer, setAiDrawer] = useState<RuntimeAiDrawerConfig | null>(null);
  const appLayoutMessageHandler = useStableCallback((event: AppLayoutMessage) => {
    if (event.type === 'registerLeftDrawer') {
      setAiDrawer(event.payload);
      if (!aiDrawerWasOpenRef.current && event.payload.defaultActive) {
        onAiDrawersChangeStable(event.payload.id, { initiatedByUserAction: false });
      }
      return;
    }
    if (aiDrawer && aiDrawer.id !== event.payload.id) {
      metrics.sendOpsMetricObject('awsui-widget-drawer-incorrect-id', { oldId: aiDrawer?.id, newId: event.payload.id });
      return;
    }
    switch (event.type) {
      case 'updateDrawerConfig':
        setAiDrawer(current => (current ? { ...current, ...event.payload } : current));
        break;
      case 'openDrawer':
        onActiveAiDrawerChangeStable(event.payload.id, { initiatedByUserAction: false });
        break;
      case 'closeDrawer':
        onActiveAiDrawerChangeStable(null, { initiatedByUserAction: false });
        break;
      case 'resizeDrawer':
        onActiveAiDrawerResizeStable(event.payload.size);
        break;
      /* istanbul ignore next: this code is not intended to be visited */
      default:
        assertNever(event);
    }
  });
  const onAiDrawersChangeStable = useStableCallback(onActiveAiDrawerChange);
  const onActiveAiDrawerResizeStable = useStableCallback(onActiveAiDrawerResize);
  const onActiveAiDrawerChangeStable = useStableCallback(onActiveAiDrawerChange);
  const aiDrawerWasOpenRef = useRef(false);
  aiDrawerWasOpenRef.current = aiDrawerWasOpenRef.current || !!activeAiDrawerId;

  useEffect(() => {
    if (!isEnabled) {
      return;
    }

    const initialDrawerMessage = getAppLayoutInitialState()?.find(message => message.type === 'registerLeftDrawer');
    if (initialDrawerMessage && initialDrawerMessage.type === 'registerLeftDrawer') {
      setAiDrawer(initialDrawerMessage.payload);
      if (!aiDrawerWasOpenRef.current && initialDrawerMessage.payload.defaultActive) {
        onAiDrawersChangeStable(initialDrawerMessage.payload.id, { initiatedByUserAction: false });
      }
    }

    const unsubscribe = registerAppLayoutHandler(appLayoutMessageHandler);
    return () => {
      unsubscribe();
      setAiDrawer(null);
    };
  }, [isEnabled, appLayoutMessageHandler, onAiDrawersChangeStable, onActiveAiDrawerResizeStable]);

  return aiDrawer && mapRuntimeConfigToAiDrawer(aiDrawer);
}

export const MIN_DRAWER_SIZE = 290;

interface UseDrawersProps {
  isEnabled: boolean;
  onAiDrawerFocus: () => void;
  expandedDrawerId: string | null;
  setExpandedDrawerId: (value: string | null) => void;
}

export function useAiDrawer({ isEnabled, onAiDrawerFocus, expandedDrawerId, setExpandedDrawerId }: UseDrawersProps) {
  const [activeAiDrawerId, setActiveAiDrawerId] = useState<string | null>(null);
  const [size, setSize] = useState<number | null>(null);

  function onActiveAiDrawerResize(size: number) {
    setSize(size);
    fireNonCancelableEvent(activeAiDrawer?.onResize, { id: activeAiDrawerId, size });
  }

  function onActiveAiDrawerChange(
    newDrawerId: string | null,
    { initiatedByUserAction }: OnChangeParams = DEFAULT_ON_CHANGE_PARAMS
  ) {
    setActiveAiDrawerId(newDrawerId);

    if (newDrawerId) {
      fireNonCancelableEvent(aiDrawer?.onToggle, { isOpen: true, initiatedByUserAction });
    }

    if (activeAiDrawerId) {
      fireNonCancelableEvent(aiDrawer?.onToggle, { isOpen: false, initiatedByUserAction });

      if (activeAiDrawerId === expandedDrawerId) {
        setExpandedDrawerId?.(null);
      }
    }

    onAiDrawerFocus?.();
  }

  const aiDrawer = useRuntimeAiDrawer(isEnabled, activeAiDrawerId, onActiveAiDrawerChange, onActiveAiDrawerResize);
  const activeAiDrawer = activeAiDrawerId && activeAiDrawerId === aiDrawer?.id ? aiDrawer : null;
  const activeAiDrawerSize = activeAiDrawerId ? (size ?? activeAiDrawer?.defaultSize ?? MIN_DRAWER_SIZE) : 0;
  const minAiDrawerSize = Math.min(activeAiDrawer?.defaultSize ?? MIN_DRAWER_SIZE, MIN_DRAWER_SIZE);

  return {
    aiDrawer,
    onActiveAiDrawerChange,
    activeAiDrawer,
    activeAiDrawerId,
    activeAiDrawerSize,
    minAiDrawerSize,
    onActiveAiDrawerResize,
  };
}
