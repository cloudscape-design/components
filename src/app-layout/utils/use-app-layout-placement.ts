// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
import { useCallback, useEffect, useRef, useState } from 'react';
import { useObservedElement } from './use-observed-element';
import { useResizeObserver } from '@cloudscape-design/component-toolkit/internal';
import { getLogicalBoundingClientRect } from '../../internal/direction';

export function useAppLayoutPlacement(headerSelector: string, footerSelector: string) {
  const mainElementRef = useRef<HTMLElement>(null);
  const headerHeight = useObservedElement(headerSelector);
  const footerHeight = useObservedElement(footerSelector);
  const [offsets, setOffsets] = useState({
    insetInlineStart: 0,
    insetInlineEnd: 0,
    inlineSize: Number.POSITIVE_INFINITY,
  });

  const updatePosition = useCallback(() => {
    if (!mainElementRef.current) {
      return;
    }
    const { insetInlineStart, insetInlineEnd, inlineSize } = getLogicalBoundingClientRect(mainElementRef.current);

    // skip reading sizes in JSDOM
    if (inlineSize === 0) {
      return;
    }

    setOffsets({ insetInlineStart, insetInlineEnd: inlineSize - insetInlineEnd, inlineSize });
  }, []);

  useEffect(() => {
    window.addEventListener('resize', updatePosition);
    return () => window.removeEventListener('resize', updatePosition);
  }, [updatePosition]);

  useResizeObserver(mainElementRef, updatePosition);

  return [mainElementRef, { ...offsets, insetBlockStart: headerHeight, insetBlockEnd: footerHeight }] as const;
}
