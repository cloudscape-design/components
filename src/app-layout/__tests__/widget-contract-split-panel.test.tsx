// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
/* eslint-disable simple-import-sort/imports */
import React from 'react';
import { render, waitFor } from '@testing-library/react';

import { describeEachAppLayout } from './utils';
import AppLayout from '../../../lib/components/app-layout';
import SplitPanel from '../../../lib/components/split-panel';
import createWrapper from '../../../lib/components/test-utils/selectors';

const isObject = (value: any) => Object.prototype.toString.call(value) === '[object Object]';

function sanitizeProps(props: any): any {
  if (!isObject(props)) {
    return props;
  }
  if (React.isValidElement(props) || props?.current instanceof Element) {
    return '__JSX__';
  }
  return Object.fromEntries(
    Object.entries(props).map(([key, value]) => {
      return [key, sanitizeProps(value)];
    })
  );
}

const renderedProps = new Map();
function createWidgetizedComponentMock(Implementation: React.ComponentType) {
  return () => {
    return function Widgetized(props: any) {
      renderedProps.set(Implementation, sanitizeProps(props));
      return <Implementation {...(props as any)} />;
    };
  };
}

jest.mock('../../../lib/components/internal/widgets', () => ({
  createWidgetizedComponent: createWidgetizedComponentMock,
}));
jest.mock('@cloudscape-design/component-toolkit/internal', () => {
  let counter = 0;
  return {
    ...jest.requireActual('@cloudscape-design/component-toolkit/internal'),
    useUniqueId: (prefix: string) => `${prefix}${++counter}`,
    useRandomId: (prefix: string) => `${prefix}${++counter}`,
  };
});

describeEachAppLayout({ themes: ['refresh-toolbar'], sizes: ['desktop'] }, () => {
  beforeEach(() => {
    renderedProps.clear();
  });

  test('contract with split panel', async () => {
    render(
      <AppLayout
        splitPanel={
          <SplitPanel header="Split panel header">
            <div>split panel content</div>
          </SplitPanel>
        }
      />
    );
    await waitFor(() => {
      expect(createWrapper().findSplitPanel()).toBeTruthy();
    });
    expect(renderedProps).toMatchSnapshot();
  });

  test('contract with split panel (trigger is hidden)', async () => {
    render(
      <AppLayout
        splitPanel={
          <SplitPanel header="Split panel header" closeBehavior="hide">
            <div>split panel content</div>
          </SplitPanel>
        }
      />
    );
    await waitFor(() => {
      expect(createWrapper().findSplitPanel()).toBeTruthy();
    });
    expect(renderedProps).toMatchSnapshot();
  });
});
