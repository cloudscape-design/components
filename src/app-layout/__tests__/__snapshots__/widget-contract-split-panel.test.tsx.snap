// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Theme=refresh-toolbar, Size=desktop contract with split panel (trigger is hidden) 1`] = `
Map {
  "BeforeMainSlotImplementation" => {
    "appLayoutProps": {
      "ariaLabels": {
        "drawers": undefined,
        "drawersOverflow": undefined,
        "drawersOverflowWithBadge": undefined,
        "navigation": undefined,
        "navigationClose": undefined,
        "navigationToggle": undefined,
        "notifications": undefined,
        "tools": undefined,
        "toolsClose": undefined,
        "toolsToggle": undefined,
      },
      "contentType": "default",
      "maxContentWidth": undefined,
      "minContentWidth": 280,
      "navigationOpen": true,
      "navigationWidth": 280,
      "onNavigationChange": [Function],
      "placement": {
        "inlineSize": Infinity,
        "insetBlockEnd": 0,
        "insetBlockStart": 0,
        "insetInlineEnd": 0,
        "insetInlineStart": 0,
      },
      "splitPanel": <SplitPanel
        closeBehavior="hide"
        header="Split panel header"
      >
        <div>
          split panel content
        </div>
      </SplitPanel>,
      "toolsWidth": 290,
    },
    "appLayoutState": {
      "isIntersecting": true,
    },
    "toolbarProps": {
      "ariaLabels": {
        "drawers": undefined,
        "drawersOverflow": undefined,
        "drawersOverflowWithBadge": undefined,
        "navigation": undefined,
        "navigationClose": undefined,
        "navigationToggle": undefined,
        "notifications": undefined,
        "tools": undefined,
        "toolsClose": undefined,
        "toolsToggle": undefined,
      },
      "hasNavigation": true,
      "navigationFocusRef": undefined,
      "navigationOpen": true,
      "onNavigationToggle": undefined,
    },
  },
  "TopContentSlotImplementation" => {
    "appLayoutProps": {
      "ariaLabels": {
        "drawers": undefined,
        "drawersOverflow": undefined,
        "drawersOverflowWithBadge": undefined,
        "navigation": undefined,
        "navigationClose": undefined,
        "navigationToggle": undefined,
        "notifications": undefined,
        "tools": undefined,
        "toolsClose": undefined,
        "toolsToggle": undefined,
      },
      "contentType": "default",
      "maxContentWidth": undefined,
      "minContentWidth": 280,
      "navigationOpen": true,
      "navigationWidth": 280,
      "onNavigationChange": [Function],
      "placement": {
        "inlineSize": Infinity,
        "insetBlockEnd": 0,
        "insetBlockStart": 0,
        "insetInlineEnd": 0,
        "insetInlineStart": 0,
      },
      "splitPanel": <SplitPanel
        closeBehavior="hide"
        header="Split panel header"
      >
        <div>
          split panel content
        </div>
      </SplitPanel>,
      "toolsWidth": 290,
    },
    "appLayoutState": {
      "isIntersecting": true,
    },
    "toolbarProps": {
      "ariaLabels": {
        "drawers": undefined,
        "drawersOverflow": undefined,
        "drawersOverflowWithBadge": undefined,
        "navigation": undefined,
        "navigationClose": undefined,
        "navigationToggle": undefined,
        "notifications": undefined,
        "tools": undefined,
        "toolsClose": undefined,
        "toolsToggle": undefined,
      },
      "hasNavigation": true,
      "navigationFocusRef": undefined,
      "navigationOpen": true,
      "onNavigationToggle": undefined,
    },
  },
  "BottomContentSlotImplementation" => {
    "appLayoutProps": {
      "ariaLabels": {
        "drawers": undefined,
        "drawersOverflow": undefined,
        "drawersOverflowWithBadge": undefined,
        "navigation": undefined,
        "navigationClose": undefined,
        "navigationToggle": undefined,
        "notifications": undefined,
        "tools": undefined,
        "toolsClose": undefined,
        "toolsToggle": undefined,
      },
      "contentType": "default",
      "maxContentWidth": undefined,
      "minContentWidth": 280,
      "navigationOpen": true,
      "navigationWidth": 280,
      "onNavigationChange": [Function],
      "placement": {
        "inlineSize": Infinity,
        "insetBlockEnd": 0,
        "insetBlockStart": 0,
        "insetInlineEnd": 0,
        "insetInlineStart": 0,
      },
      "splitPanel": <SplitPanel
        closeBehavior="hide"
        header="Split panel header"
      >
        <div>
          split panel content
        </div>
      </SplitPanel>,
      "toolsWidth": 290,
    },
    "appLayoutState": {
      "isIntersecting": true,
    },
    "toolbarProps": {
      "ariaLabels": {
        "drawers": undefined,
        "drawersOverflow": undefined,
        "drawersOverflowWithBadge": undefined,
        "navigation": undefined,
        "navigationClose": undefined,
        "navigationToggle": undefined,
        "notifications": undefined,
        "tools": undefined,
        "toolsClose": undefined,
        "toolsToggle": undefined,
      },
      "hasNavigation": true,
      "navigationFocusRef": undefined,
      "navigationOpen": true,
      "onNavigationToggle": undefined,
    },
  },
  "AfterMainSlotImplementation" => {
    "appLayoutProps": {
      "ariaLabels": {
        "drawers": undefined,
        "drawersOverflow": undefined,
        "drawersOverflowWithBadge": undefined,
        "navigation": undefined,
        "navigationClose": undefined,
        "navigationToggle": undefined,
        "notifications": undefined,
        "tools": undefined,
        "toolsClose": undefined,
        "toolsToggle": undefined,
      },
      "contentType": "default",
      "maxContentWidth": undefined,
      "minContentWidth": 280,
      "navigationOpen": true,
      "navigationWidth": 280,
      "onNavigationChange": [Function],
      "placement": {
        "inlineSize": Infinity,
        "insetBlockEnd": 0,
        "insetBlockStart": 0,
        "insetInlineEnd": 0,
        "insetInlineStart": 0,
      },
      "splitPanel": <SplitPanel
        closeBehavior="hide"
        header="Split panel header"
      >
        <div>
          split panel content
        </div>
      </SplitPanel>,
      "toolsWidth": 290,
    },
    "appLayoutState": {
      "isIntersecting": true,
    },
    "toolbarProps": {
      "ariaLabels": {
        "drawers": undefined,
        "drawersOverflow": undefined,
        "drawersOverflowWithBadge": undefined,
        "navigation": undefined,
        "navigationClose": undefined,
        "navigationToggle": undefined,
        "notifications": undefined,
        "tools": undefined,
        "toolsClose": undefined,
        "toolsToggle": undefined,
      },
      "hasNavigation": true,
      "navigationFocusRef": undefined,
      "navigationOpen": true,
      "onNavigationToggle": undefined,
    },
  },
  "AppLayoutStateProvider" => {
    "appLayoutProps": {
      "ariaLabels": {
        "drawers": undefined,
        "drawersOverflow": undefined,
        "drawersOverflowWithBadge": undefined,
        "navigation": undefined,
        "navigationClose": undefined,
        "navigationToggle": undefined,
        "notifications": undefined,
        "tools": undefined,
        "toolsClose": undefined,
        "toolsToggle": undefined,
      },
      "contentType": "default",
      "maxContentWidth": undefined,
      "minContentWidth": 280,
      "navigationOpen": true,
      "navigationWidth": 280,
      "onNavigationChange": [Function],
      "placement": {
        "inlineSize": Infinity,
        "insetBlockEnd": 0,
        "insetBlockStart": 0,
        "insetInlineEnd": 0,
        "insetInlineStart": 0,
      },
      "splitPanel": <SplitPanel
        closeBehavior="hide"
        header="Split panel header"
      >
        <div>
          split panel content
        </div>
      </SplitPanel>,
      "toolsWidth": 290,
    },
    "forwardRef": null,
    "stateManager": {
      "current": {
        "hasToolbar": true,
        "setState": [Function],
        "setToolbar": undefined,
      },
    },
  },
}
`;

exports[`Theme=refresh-toolbar, Size=desktop contract with split panel 1`] = `
Map {
  "BeforeMainSlotImplementation" => {
    "appLayoutProps": {
      "ariaLabels": {
        "drawers": undefined,
        "drawersOverflow": undefined,
        "drawersOverflowWithBadge": undefined,
        "navigation": undefined,
        "navigationClose": undefined,
        "navigationToggle": undefined,
        "notifications": undefined,
        "tools": undefined,
        "toolsClose": undefined,
        "toolsToggle": undefined,
      },
      "contentType": "default",
      "maxContentWidth": undefined,
      "minContentWidth": 280,
      "navigationOpen": true,
      "navigationWidth": 280,
      "onNavigationChange": [Function],
      "placement": {
        "inlineSize": Infinity,
        "insetBlockEnd": 0,
        "insetBlockStart": 0,
        "insetInlineEnd": 0,
        "insetInlineStart": 0,
      },
      "splitPanel": <SplitPanel
        header="Split panel header"
      >
        <div>
          split panel content
        </div>
      </SplitPanel>,
      "toolsWidth": 290,
    },
    "appLayoutState": {
      "isIntersecting": true,
    },
    "toolbarProps": {
      "ariaLabels": {
        "drawers": undefined,
        "drawersOverflow": undefined,
        "drawersOverflowWithBadge": undefined,
        "navigation": undefined,
        "navigationClose": undefined,
        "navigationToggle": undefined,
        "notifications": undefined,
        "tools": undefined,
        "toolsClose": undefined,
        "toolsToggle": undefined,
      },
      "hasNavigation": true,
      "navigationFocusRef": undefined,
      "navigationOpen": true,
      "onNavigationToggle": undefined,
    },
  },
  "TopContentSlotImplementation" => {
    "appLayoutProps": {
      "ariaLabels": {
        "drawers": undefined,
        "drawersOverflow": undefined,
        "drawersOverflowWithBadge": undefined,
        "navigation": undefined,
        "navigationClose": undefined,
        "navigationToggle": undefined,
        "notifications": undefined,
        "tools": undefined,
        "toolsClose": undefined,
        "toolsToggle": undefined,
      },
      "contentType": "default",
      "maxContentWidth": undefined,
      "minContentWidth": 280,
      "navigationOpen": true,
      "navigationWidth": 280,
      "onNavigationChange": [Function],
      "placement": {
        "inlineSize": Infinity,
        "insetBlockEnd": 0,
        "insetBlockStart": 0,
        "insetInlineEnd": 0,
        "insetInlineStart": 0,
      },
      "splitPanel": <SplitPanel
        header="Split panel header"
      >
        <div>
          split panel content
        </div>
      </SplitPanel>,
      "toolsWidth": 290,
    },
    "appLayoutState": {
      "isIntersecting": true,
    },
    "toolbarProps": {
      "ariaLabels": {
        "drawers": undefined,
        "drawersOverflow": undefined,
        "drawersOverflowWithBadge": undefined,
        "navigation": undefined,
        "navigationClose": undefined,
        "navigationToggle": undefined,
        "notifications": undefined,
        "tools": undefined,
        "toolsClose": undefined,
        "toolsToggle": undefined,
      },
      "hasNavigation": true,
      "navigationFocusRef": undefined,
      "navigationOpen": true,
      "onNavigationToggle": undefined,
    },
  },
  "BottomContentSlotImplementation" => {
    "appLayoutProps": {
      "ariaLabels": {
        "drawers": undefined,
        "drawersOverflow": undefined,
        "drawersOverflowWithBadge": undefined,
        "navigation": undefined,
        "navigationClose": undefined,
        "navigationToggle": undefined,
        "notifications": undefined,
        "tools": undefined,
        "toolsClose": undefined,
        "toolsToggle": undefined,
      },
      "contentType": "default",
      "maxContentWidth": undefined,
      "minContentWidth": 280,
      "navigationOpen": true,
      "navigationWidth": 280,
      "onNavigationChange": [Function],
      "placement": {
        "inlineSize": Infinity,
        "insetBlockEnd": 0,
        "insetBlockStart": 0,
        "insetInlineEnd": 0,
        "insetInlineStart": 0,
      },
      "splitPanel": <SplitPanel
        header="Split panel header"
      >
        <div>
          split panel content
        </div>
      </SplitPanel>,
      "toolsWidth": 290,
    },
    "appLayoutState": {
      "isIntersecting": true,
    },
    "toolbarProps": {
      "ariaLabels": {
        "drawers": undefined,
        "drawersOverflow": undefined,
        "drawersOverflowWithBadge": undefined,
        "navigation": undefined,
        "navigationClose": undefined,
        "navigationToggle": undefined,
        "notifications": undefined,
        "tools": undefined,
        "toolsClose": undefined,
        "toolsToggle": undefined,
      },
      "hasNavigation": true,
      "navigationFocusRef": undefined,
      "navigationOpen": true,
      "onNavigationToggle": undefined,
    },
  },
  "AfterMainSlotImplementation" => {
    "appLayoutProps": {
      "ariaLabels": {
        "drawers": undefined,
        "drawersOverflow": undefined,
        "drawersOverflowWithBadge": undefined,
        "navigation": undefined,
        "navigationClose": undefined,
        "navigationToggle": undefined,
        "notifications": undefined,
        "tools": undefined,
        "toolsClose": undefined,
        "toolsToggle": undefined,
      },
      "contentType": "default",
      "maxContentWidth": undefined,
      "minContentWidth": 280,
      "navigationOpen": true,
      "navigationWidth": 280,
      "onNavigationChange": [Function],
      "placement": {
        "inlineSize": Infinity,
        "insetBlockEnd": 0,
        "insetBlockStart": 0,
        "insetInlineEnd": 0,
        "insetInlineStart": 0,
      },
      "splitPanel": <SplitPanel
        header="Split panel header"
      >
        <div>
          split panel content
        </div>
      </SplitPanel>,
      "toolsWidth": 290,
    },
    "appLayoutState": {
      "isIntersecting": true,
    },
    "toolbarProps": {
      "ariaLabels": {
        "drawers": undefined,
        "drawersOverflow": undefined,
        "drawersOverflowWithBadge": undefined,
        "navigation": undefined,
        "navigationClose": undefined,
        "navigationToggle": undefined,
        "notifications": undefined,
        "tools": undefined,
        "toolsClose": undefined,
        "toolsToggle": undefined,
      },
      "hasNavigation": true,
      "navigationFocusRef": undefined,
      "navigationOpen": true,
      "onNavigationToggle": undefined,
    },
  },
  "AppLayoutStateProvider" => {
    "appLayoutProps": {
      "ariaLabels": {
        "drawers": undefined,
        "drawersOverflow": undefined,
        "drawersOverflowWithBadge": undefined,
        "navigation": undefined,
        "navigationClose": undefined,
        "navigationToggle": undefined,
        "notifications": undefined,
        "tools": undefined,
        "toolsClose": undefined,
        "toolsToggle": undefined,
      },
      "contentType": "default",
      "maxContentWidth": undefined,
      "minContentWidth": 280,
      "navigationOpen": true,
      "navigationWidth": 280,
      "onNavigationChange": [Function],
      "placement": {
        "inlineSize": Infinity,
        "insetBlockEnd": 0,
        "insetBlockStart": 0,
        "insetInlineEnd": 0,
        "insetInlineStart": 0,
      },
      "splitPanel": <SplitPanel
        header="Split panel header"
      >
        <div>
          split panel content
        </div>
      </SplitPanel>,
      "toolsWidth": 290,
    },
    "forwardRef": null,
    "stateManager": {
      "current": {
        "hasToolbar": true,
        "setState": [Function],
        "setToolbar": undefined,
      },
    },
  },
}
`;
